### ISSUE TITLE
Control Widgets Tab Focus Order Via Code

### ISSUE THREAD

**Issue by saroad2** [beeware/toga#1650]
Is your feature request related to a problem? Please describe.
========================================

*TL;DR - one should be able to control the tab focus order of widgets in a window.*

Assume you have a window with two buttons in a line named `button1` and `button2`.

If you insert `button3` between those two buttons, the tab order of the buttons will be the **order of insertion**, i.e. `button1`, `button2`, `button3`. You might want to change that order to be the **order of appearance on the screen**, i.e. `button1`, `button3`, `button2`.

This is a not bug, in my opinon. However, we would like to let our users have a way to control the tab order manually via code.

Describe the solution you'd like
====================
The way I see it, I can think of 3 possible APIs for that abbility:

Option I - `widget.next_tab` and `widget.prev_tab`
------------------------------------------------------
With this API, for each widget one should be able to define the next tab or prev tab. For example:
```python
widget1.next_tab = widget2
widget1.prev_tab = widget3
```

The problem with this approach is that it introduces a way in which one can jump between windows if we set the next tab of a widget to be a widget from another window. We can simply raise an exception if someone does that.

Option II - `widget.tab_index`
--------------------------------
another approach should be to set the tab index of a widget. In that way, the tab order will be selected for each window by the order of tab index. This is similar to the way the tab order is handled in [.Net](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.control.tabindex?view=windowsdesktop-6.0). For example:
```python
widget1.tab_index = 0
widget2.tab_index = 10
widget3.tab_index = 100
```

The problem with this approach that we have to keep track of the index of other widgets in order to decied which index to use for our widget.

Another problem happens in the case where we choose the tab index to be an index we already chose for another widget. We can solve this problem by bumping by 1 the index for every widget with our index or above.

Option III - `window.set_tab_index(widget, index)`
------------------------------------------------------
Another approach is to control the widgets order from the `window` class. In that way we can make sure that we do not set the index of a widget to be the same as another widget and we can make sure that we order the tab index of widgets which are only from our window.

The downside of this approach is that we cannot simply set the "next widget" or "previous widget" , we have to do this using the `window` instance.

Summary
============
I don't have a favorite approach, but I tend towards option II. We can also choose to implement more than one of those APIs.

I think this is an important feature and I can think of many use cases where one might need it.
Let me know what you think.

**Comment by freakboy3742** [comment]
Agreed that this is a problem for which we need a solution (although it's not high on my list of priorities at the moment given the other gaps that exist in Toga at present).

On first impressions, I think I agree that option II makes the most sense; it has the added benefit of [matching the general approach exposed by HTML for focus control](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex), which is always a good sign. However, I wouldn't want to commit to anything without doing more research on the native APIs exposed by other platforms in case Cocoa/UIKit/GTK/etc exposes radically different APIs. That doesn't mean we necessarily need to have an implementation for every platform before this would be accepted into core; I just want to make sure we don't paint ourselves into a corner.

**Comment by saroad2** [comment]
Hey, thanks for the reply.

I understand why this is a low priority right now, although users of Eddington did complain about it.

I did a little checkup and it seems that [Cocoa is following option I](https://developer.apple.com/documentation/appkit/nsview/1483465-nextkeyview). I couldn't find the answer for GTK and the others.

Once we agree on the right approach, I'll go ahead and implement it.

**Comment by freakboy3742** [comment]
To be clear - it's a low priority *for me*, which only impacts whether I'm likely to start working on it myself. If you've got users who want this feature, and you're motivated to work on it, then I'm happy to review any submitted PR.

A quick look at the GTK docs seems to suggest that they have 2 approaches - a deprecated approach which is well documented and understood, most closely matching option 1; and a new approach which is neither documented nor understood.

Thinking more generally - regardless of which API approach we take, we should be able to accomodate the other API. An index-based approach can be converted to a direct sequence by enumerating the sequence and applying the required focus chain; a direct sequence can be converted into an index based approach by assigning the position in the chain as the index. The index-based approach will be *marginally* less prone to error, as "closed loops" (.e.g, widget 1->widget 2->widget 1, missing widget 3 in the tab sequence) can't be accidentally constructed; so I think I'm inclined to go with the index-based approach.

**Comment by bruno-rino** [comment]
Having no skin in the game, I'd like to make sure we settle for the best API. I started this discussion in the [PR](https://github.com/beeware/toga/pull/1686), but this is indeed the right place for it.

Looking at it from "what feels more natural" standpoint, I'd rather see a different API: `window.set_focus_chain(list_of_widgets)`, like the (deprecated) [gtk API](https://docs.gtk.org/gtk3/method.Container.set_focus_chain.html). 

Advantages:
- does away with indexes, which feels like an unnecessary abstraction. 
- avoids the question of the scope of the indexes. Is it the window, or the parent widget? On winforms, _"Container controls, such as a [GroupBox](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.groupbox) control, treat their children as separate from the rest of the form."_ (from [here](https://learn.microsoft.com/en-us/dotnet/desktop/winforms/controls/how-to-set-the-tab-order?view=netdesktop-6.0))

Drawbacks to a `focus_chain` API:
- as mentioned, one can accidentally forget to add a widget to the list
- if a widget is dynamically added, one has to insert it in the list at the desired location, which is not as trivial as using indexes.

I find those drawbacks pretty minor. Avoiding an unnecessary abstraction is better.

---

If the "tab_index" API is chosen:

- there's the question of naming: I'd rather see "focus_index", which is more conceptual.
- should the `tab_stop` property, to "remove a control from the tab order",  be implemented as well?

**Comment by saroad2** [comment]
I'm reopening the issue since it was not fixed for cocoa and gtk platforms

**Comment by johnzhou721** [comment]
Also not implemented for Qt, Web, or Textual, for that matter.
